diff --git a/Makefile b/Makefile
index d66c156be1c6..40f30653c49b 100644
--- a/Makefile
+++ b/Makefile
@@ -874,6 +874,12 @@ else
 KBUILD_CFLAGS	+= -mllvm -polly-opt-fusion=max
 endif
 
+KBUILD_CFLAGS += -fclang-abi-compat=18.0.0
+KBUILD_CFLAGS += -fno-addrsig
+KBUILD_CFLAGS += -fno-unique-section-names
+KBUILD_CFLAGS += -mllvm -disable-peephole
+KBUILD_CFLAGS += -fno-semantic-interposition
+
 # Polly may optimise loops with dead paths beyound what the linker
 # can understand. This may negate the effect of the linker's DCE
 # so we tell Polly to perfom proven DCE on the loops it optimises
