diff --git a/Makefile b/Makefile
index d66c156be1c6..6a952b231eb3 100644
--- a/Makefile
+++ b/Makefile
@@ -2090,6 +2090,20 @@ clang-tidy clang-analyzer:
 	@false
 endif
 
+ifdef CONFIG_CC_IS_CLANG
+# 结构体类型稳定性
+KBUILD_CFLAGS += -fno-unique-section-names \
+                 -fno-eliminate-unused-debug-types
+
+# 符号CRC保护
+KBUILD_CFLAGS += -mllvm -disable-auto-upgrade-debug-info \
+                 -mllvm -generate-type-units
+
+# 优化行为控制
+KBUILD_CFLAGS += -fno-strict-aliasing \
+                 -fno-optimize-sibling-calls
+endif
+
 # Scripts to check various things for consistency
 # ---------------------------------------------------------------------------
 
